### PR TITLE
removed sdb_fmt in /arch/

### PR DESCRIPTION
**Pull request by HN026**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

...

**Test plan**

...

**Closing issues**

Partially addresses #1168 
...

### CODE REVIEW

**Comment by p371jah** [comment]
IMO allocating strings on heap is not the best choice.

**Comment by HN026** [comment]
> IMO allocating strings on heap is not the best choice.

Other alternatives come w their trade-offs

File: librz/asm/arch/6502/6502dis.c [hunk 1]
```diff
@@ -143,40 +143,49 @@ static int _6502Disass(ut64 pc, RzAsmOp *op, const ut8 *buf, ut64 len) {
 	int i;
 	for (i = 0; ops[i].name != NULL; i++) {
 		if (ops[i].op == buf[0]) {
-			const char *buf_asm = "invalid";
+			char *buf_asm = rz_str_newf("invalid");
```
Comment (wargio) [inline librz/asm/arch/6502/6502dis.c:146]: You could start as `NULL` and if it stays NULL before the end then you can set this as `rz_str_new("invalid")`

File: librz/asm/arch/6502/6502dis.c [hunk 2]
```diff
@@ -143,40 +143,49 @@ static int _6502Disass(ut64 pc, RzAsmOp *op, const ut8 *buf, ut64 len) {
...
 				break;
 			case 3:
 				if (len > 2) {
-					buf_asm = sdb_fmt(ops[i].name, buf[1] + 0x100 * buf[2]);
+					free(buf_asm);
+					buf_asm = rz_str_newf(ops[i].name, buf[1] + 0x100 * buf[2]);
 				} else {
-					buf_asm = "truncated";
+					free(buf_asm);
+					buf_asm = rz_str_newf("truncated");
```
Comment (wargio) [inline librz/asm/arch/6502/6502dis.c]: use `rz_str_newf` only for those strings that requires formatting (`"foo %d %s..."`) and for strings like these instead just `rz_str_new`

File: librz/asm/arch/6502/6502dis.c (hunk 1 above)
Comment (wargio) [inline librz/asm/arch/6502/6502dis.c]: set this to NULL.

File: librz/asm/arch/6502/6502dis.c [hunk 3]
```diff
@@ -143,40 +143,49 @@ static int _6502Disass(ut64 pc, RzAsmOp *op, const ut8 *buf, ut64 len) {
...
+					buf_asm = rz_str_newf(ops[i].name, buf[1] + 0x100 * buf[2] + 0x10000 * buf[3]);
 				} else {
-					buf_asm = "truncated";
+					free(buf_asm);
+					buf_asm = rz_str_newf("truncated");
 					len = -1;
 				}
 				break;
 			default:
+				free(buf_asm);
```
Comment (wargio) [inline librz/asm/arch/6502/6502dis.c]:
```suggestion
				buf_asm = rz_str_new("invalid");
```

File: librz/asm/arch/6502/6502dis.c (hunk 1 above)
Comment (wargio) [reply librz/asm/arch/6502/6502dis.c]: and remove all the `free` calls you have added but not the one after `rz_strbuf_set`

File: librz/asm/arch/gb/gbdis.c [hunk 4]
```diff
@@ -156,28 +156,34 @@ static int gbDisass(RzAsmOp *op, const ut8 *buf, int len) {
...
+		free(buf_asm);
+		buf_asm = rz_str_newf(gb_op[buf[0]].name, buf[1] + 0x100 * buf[2]);
 		break;
 	case GB_8BIT + ARG_8 + GB_IO:
 		gb_hardware_register_name(reg, buf[1]);
-		buf_asm = sdb_fmt(gb_op[buf[0]].name, reg);
+		free(buf_asm);
+		buf_asm = rz_str_newf(gb_op[buf[0]].name, reg);
 		break;
 	}
```
Comment (wargio) [inline librz/asm/arch/gb/gbdis.c:181]: add a `default` case and use that to set the `invalid` string.

File: librz/asm/arch/i4004/i4004dis.c [hunk 5]
```diff
@@ -62,40 +62,86 @@ static int i4004dis(RzAsmOp *op, const ut8 *buf, int len) {
 	int rlen = i4004_get_ins_len(*buf);
 	ut8 high = (*buf & 0xf0) >> 4;
 	ut8 low = (*buf & 0xf);
-	const char *buf_asm = "invalid";
+	char *buf_asm = rz_str_newf("invalid");
```
Comment (wargio) [inline librz/asm/arch/i4004/i4004dis.c]: same here. add a `default` case, set the invalid there and remove the added `free` but not the one after the `rz_strbuf_set`

**Comment by wargio** [comment]
> IMO allocating strings on heap is not the best choice.

true but `sdb_fmt` uses an internal `static char buffer[SIZE]` which forbids us to do multithreading.

**Comment by wargio** [comment]
@pelijah i have added this change so @HN026 can use this instead of allocating mem https://github.com/rizinorg/rizin/pull/4124

**Comment by wargio** [comment]
@HN026 Please rebase and use the new `setf_asm` function

File: librz/asm/arch/6502/6502dis.c [hunk 6]
```diff
@@ -143,52 +143,42 @@ static int _6502Disass(ut64 pc, RzAsmOp *op, const ut8 *buf, ut64 len) {
...
-				free(buf_asm);
 				goto beach;
 			}
-			rz_strbuf_set(&op->buf_asm, buf_asm);
-			free(buf_asm);
 			return len;
 		}
 	}
 beach:
+	rz_asm_op_set_asm(op, "invalid");
```
Comment (wargio) [inline librz/asm/arch/6502/6502dis.c]: no. this overwrites the other `rz_asm_op_set_asm`. move this into the `default` case. (see the other comment.)

File: librz/asm/arch/6502/6502dis.c [hunk 7]
```diff
@@ -143,52 +143,42 @@ static int _6502Disass(ut64 pc, RzAsmOp *op, const ut8 *buf, ut64 len) {
...
+					rz_asm_op_setf_asm(op, ops[i].name, buf[1] + 0x100 * buf[2] + 0x10000 * buf[3]);
 				} else {
-					free(buf_asm);
-					buf_asm = rz_str_newf("truncated");
+					rz_asm_op_set_asm(op, "truncated");
 					len = -1;
 				}
 				break;
 			default:
-				free(buf_asm);
```
Comment (wargio) [inline librz/asm/arch/6502/6502dis.c]: Move here the `rz_asm_op_set_asm(op, "invalid");`

**Review by wargio — CHANGES_REQUESTED** [review]
Looks almost perfect besides for the issue in `librz/asm/arch/6502/6502dis.c`

**Review by XVilka — APPROVED** [review]

**Pull request merged**